### PR TITLE
Extract CacheBufferedIndexInput from CacheDirectory

### PR DESCRIPTION
**Pull request by tlrx**
Following #53860, this pull request extracts the `CacheBufferedIndexInput` class from the `CacheDirectory` so that it can be merged with `SearchableSnapshotDirectory`.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class CacheBufferedIndexInput extends BaseSearchableSnapshotIndexInput {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java]: This class has been extracted from CacheDirectory as is, with the following modification:
- logger moved to the base class (previously at the CacheDirectory level) 
- COPY_BUFFER_SIZE moved within the CacheBufferedIndexInput class (previously at the CacheDirectory level) 
- CacheFileReference now explicitely calls directory.createCacheKey() and directory.getCacheFile()

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+                                    IndexInputStats stats, long offset, long length, boolean isClone,
+                                    @Nullable CacheFileReference cacheFileReference) {
+        super(resourceDesc, directory.blobContainer(), fileInfo, context);
+        this.directory = directory;
+        this.offset = offset;
+        this.stats = stats;
+        this.end = offset + length;
+        this.closed = new AtomicBoolean(false);
+        this.isClone = isClone;
+        this.cacheFileReference = Objects.requireNonNullElseGet(cacheFileReference,
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java]: Maybe make `CacheFileReference` a `static class` (top-level even) and pass the `directory` to its constructor? That would avoid needing to do this.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a couple of suggestions.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java [hunk 3]
```diff
@@ -19,6 +21,8 @@
 
 public abstract class BaseSearchableSnapshotIndexInput extends BufferedIndexInput {
 
+    protected static final Logger logger = LogManager.getLogger(BaseSearchableSnapshotIndexInput.class);
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java]: Makes sense

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java]: I must admit I hesitated between the two. I applied your opinion.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 4]
```diff
@@ -48,6 +50,8 @@
  */
 public class SearchableSnapshotIndexInput extends BaseSearchableSnapshotIndexInput {
 
+    private static final Logger logger = LogManager.getLogger(SearchableSnapshotIndexInput.class);
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: This isn't used? I mean it probably will be in future, but I think I'd prefer to add it when it's needed.

**Review by DaveCTurner — APPROVED** [review]
LGTM; one nit about an unused field.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 4 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: :man_facepalming: Thanks, I removed it.

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**